### PR TITLE
FreeBSD: avoid memory allocation in arc_prune_async

### PR DESCRIPTION
**Pull request by mjguzik**
Signed-off-by: Mateusz Guzik <mjguzik@gmail.com>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by freqlabs — APPROVED** [review]

**Review by behlendorf — APPROVED** [review]

File: module/os/freebsd/zfs/arc_os.c [hunk 1]
```diff
@@ -158,10 +158,13 @@ arc_default_max(uint64_t min, uint64_t allmem)
 static void
 arc_prune_task(void *arg)
 {
-	int64_t nr_scan = *(int64_t *)arg;
+#ifdef __LP64__
+	int64_t nr_scan = (int64_t)arg;
+#else
+	int64_t nr_scan = (int32_t)arg;
+#endif
```
Comment (amotin) [inline module/os/freebsd/zfs/arc_os.c:165]: Why not uintptr_t?

File: module/os/freebsd/zfs/arc_os.c [hunk 2]
```diff
@@ -185,13 +188,14 @@ arc_prune_task(void *arg)
 void
 arc_prune_async(int64_t adjust)
 {
-
 	int64_t *adjustptr;
```
Comment (amotin) [inline module/os/freebsd/zfs/arc_os.c:191]: I'd make it void * or remove completely.

File: module/os/freebsd/zfs/arc_os.c [hunk 3]
```diff
@@ -185,13 +188,14 @@ arc_prune_task(void *arg)
...
 
-	if ((adjustptr = malloc(sizeof (int64_t), M_TEMP, M_NOWAIT)) == NULL)
-		return;
+#ifndef __LP64__
+	if (adjust > __LONG_MAX)
+		adjust = __LONG_MAX;
+#endif
 
-	*adjustptr = adjust;
+	adjustptr = (void *)adjust;
```
Comment (amotin) [inline module/os/freebsd/zfs/arc_os.c:198]: I think it would be better be brought to uintptr_t first.

**Review by amotin — COMMENTED** [review]
I have no objections about the idea, but have few about implementation.

**Comment by behlendorf** [comment]
@mjguzik when you get a chance can you address the review feedback in this PR.

**Comment by behlendorf** [comment]
@tonynguien let's go ahead and move forward with the updated version of this in PR #12049.

**Comment by amotin** [comment]
Mateusz just today answered me in private email that he does not care which of the versions is committed.

**Pull request closed without merging**